modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/DynamicEnableIndexingConcurrentSelfTest.java

```diff
@@ -0,0 +1,629 @@
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.managers.discovery.CustomEventListener;
+import org.apache.ignite.internal.processors.affinity.AffinityTopologyVersion;
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.processors.query.GridQueryProcessor;
+import org.apache.ignite.internal.processors.query.h2.IgniteH2Indexing;
+import org.apache.ignite.internal.processors.query.schema.SchemaOperationException;
+import org.apache.ignite.internal.processors.query.schema.message.SchemaFinishDiscoveryMessage;
+import org.apache.ignite.internal.util.typedef.G;
+import org.apache.ignite.internal.util.typedef.T2;
+import org.apache.ignite.internal.util.typedef.X;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.transactions.TransactionSerializationException;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * Test different scnerarions on concurrent enabling indexing.
+ */
+@RunWith(Parameterized.class)
+public class DynamicEnableIndexingConcurrentSelfTest extends DynamicEnableIndexingAbstractTest {
+    /** Test parameters. */
+    @Parameters(name = "cacheMode={0},atomicityMode={1}")
+    public static Iterable<Object[]> params() {
+        CacheMode[] cacheModes = new CacheMode[] {CacheMode.PARTITIONED, CacheMode.REPLICATED};
+
+        CacheAtomicityMode[] atomicityModes = new CacheAtomicityMode[] {
+                CacheAtomicityMode.ATOMIC,
+                CacheAtomicityMode.TRANSACTIONAL,
+                CacheAtomicityMode.TRANSACTIONAL_SNAPSHOT
+        };
+
+        List<Object[]> res = new ArrayList<>();
+        for (CacheMode cacheMode : cacheModes) {
+            for (CacheAtomicityMode atomicityMode : atomicityModes)
+                res.add(new Object[] {cacheMode, atomicityMode});
+        }
+
+        return res;
+    }
+
+    /** Latches to block certain index operations. */
+    private static final ConcurrentHashMap<UUID, T2<CountDownLatch, CountDownLatch>> BLOCKS =
+            new ConcurrentHashMap<>();
+
+    /** Name field index name. */
+    private static final String NAME_FIELD_IDX_NAME = "name_idx";
+
+    /** Large number of entries. */
+    private static final int LARGE_NUM_ENTRIES = 100_000;
+
+    /** */
+    @Parameter(0)
+    public CacheMode cacheMode;
+
+    /** */
+    @Parameter(1)
+    public CacheAtomicityMode atomicityMode;
+
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        GridQueryProcessor.idxCls = null;
+
+        for (T2<CountDownLatch, CountDownLatch> block : BLOCKS.values())
```

Comment:
Why do we need this? It seems the test will pass even we removed these lines.

Reply:
Of course it will pass.